etcdserver/snapshot_store.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	dstorage "github.com/coreos/etcd/storage"
+)
+
+type snapshot struct {
+	r raftpb.Snapshot
+}
+
+func (s *snapshot) raft() raftpb.Snapshot { return s.r }
+
+func (s *snapshot) size() uint64 { return 0 }
+
+func (s *snapshot) writeTo(w io.Writer) (n int64, err error) { return 0, nil }
+
+func (s *snapshot) close() {}
+
+type snapshotStore struct {
+	// dir to save snapshot data
+	dir string
+	kv  dstorage.KV
+
+	// send empty to reqsnapc to notify the channel receiver to send back latest
+	// snapshot to snapc
+	reqsnapc chan struct{}
+	// a chan to receive the requested snapshot
+	// snapshotStore will receive from the chan immediately after it sends empty to reqsnapc
+	snapc chan raftpb.Snapshot
+
+	mu   sync.Mutex
+	snap *snapshot
+}
+
+func newSnapshotStore(dir string, kv dstorage.KV) *snapshotStore {
+	return &snapshotStore{
+		dir:      dir,
+		kv:       kv,
+		reqsnapc: make(chan struct{}),
+		snapc:    make(chan raftpb.Snapshot),
+	}
+}
+
+// getNew gets a new snapshot.
+// It returns ErrSnapshotTemporarilyUnavailable if it fails to return snapshot now.
+// TODO: the function should be unblock and finish in 1ms. It may use long time
```

Comment:
Unblock word is inaccurate. It takes some time to finish but definitely not
blocking.

On Tue, Sep 29, 2015 at 19:45 Xiang Li <notifications@github.com> wrote:

> In etcdserver/snapshot_store.go
> <https://github.com/coreos/etcd/pull/3614#discussion_r40754055>:
>
> > +	mu   sync.Mutex
> > +	snap *snapshot
> > +}
> > +
> > +func newSnapshotStore(dir string, kv dstorage.KV) *snapshotStore {
> > +	return &snapshotStore{
> > +		dir:      dir,
> > +		kv:       kv,
> > +		reqsnapc: make(chan struct{}),
> > +		snapc:    make(chan raftpb.Snapshot),
> > +	}
> > +}
> > +
> > +// getNew gets a new snapshot.
> > +// It returns ErrSnapshotTemporarilyUnavailable if it fails to return snapshot now.
> > +// TODO: the function should be unblock and finish in 1ms. It may use long time
>
> why this function is blocking right now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/3614/files#r40754055>.
>